packages/block-editor/src/components/block-list/root-container.js

```diff
@@ -79,15 +80,19 @@ function RootContainer( { children, className }, ref ) {
 		}
 	}
 
+	const [ blockNodes, setBlockNodes ] = useState( {} );
+
 	return (
 		<InsertionPoint
 			isMultiSelecting={ isMultiSelecting }
 			hasMultiSelection={ hasMultiSelection }
 			selectedBlockClientId={ selectedBlockClientId }
 			containerRef={ ref }
 		>
-			<BlockNodes.Provider value={ useState( {} ) }>
+			<BlockNodes.Provider value={ blockNodes }>
```

Comment:
It's a bit weird to use two providers for this, I think we should just do this:

```
const [ blockNodes, setBlockNodes ] = useState( {} );
const context = useMemo(() => { blockNodes, setBlockNodes }, [ blockNodes, setBlockNotes ] );
```

Reply:
If we don't use two providers, then `blockNodes` will force the block to re-render if it uses `useContext` just to be able to set the block node.

Reply:
I don't see any difference with my proposal above, I memoize the value passed to the provider which will have the same impact as if you use two providers. the value changes at the same time it does for the "two providers" approach.

Reply:
The memoized value depends on `blockNodes`. So useContext in the block wrapper will trigger a re-render of the block wrapper if `blockNodes` updates.
With two providers, `useContext` won't trigger a re-render because it's only dependent on `setState`, which stays the same throughout the component's lifecycle.

Reply:
I'm running a performance test now with `useMemo`

Reply:
I replaced the context value with his:

```
const blockNodesState = useState( {} );
const contextValue = useMemo( () => blockNodesState, blockNodesState );
```

Result:

```
Block Selection Performance:
Average time to select a block: 161.26ms
Slowest time to select a block: 214.84ms
Fastest time to select a block: 36.77ms
```

That's better than master, but not as good as this branch.

Reply:
Oh I get it now, I still don't like the double providers here though, it seems very unnatural. To be fair, I'm ok with it as a temporary measure but I'm thinking if later we'd want to access the block nodes inside the block tree itself (to replace getBlockDomNode usage for instance), we won't be able to do so.

So, this makes me wonder if we should instead try to use the store and benefit from useSelect, useDispatch.

Reply:
That's fine, we can still wrap the context provider around the blocks. I think blocks are only re-rendered anyway when the props change (or its own hooks), not necessarily when a parent re-renders.

Reply:
Can you try it? I  think you'll have the same results as the "useMemo" experiment (or I'm not understanding anaything :P)

Reply:
Yeah, I'll try it

Reply:
With 4d1c610:

```
Block Selection Performance:
Average time to select a block: 45.59ms
Slowest time to select a block: 59.66ms
Fastest time to select a block: 37.16ms
```

Reply:
So I think the main difference here is that, yes, the parent re-renders, but the blocks does not, because it's a `pure` component. if we memoize one context value, the block will suddenly re-render for every block node added by another block, because `useContext` will depend on `blockNodes`.

Reply:
mmm makes sense. I guess we should try to use the store soonish to avoid the multiplication of contexts. We already have three just for block related values, now we have two more for block nodes.

Reply:
Yeah, maybe. Tbh, I don't understand why React doesn't support multiple values on the same context.